linera-chain/src/chain.rs

```diff
@@ -641,22 +648,37 @@ where
             })
             .collect::<Vec<_>>();
         let channels = self.channels.try_load_entries_mut(&full_names).await?;
-        for ((name, id), mut channel) in raw_result.subscribe.into_iter().zip(channels) {
-            let full_name = ChannelFullName {
-                application_id,
-                name,
-            };
-            // Add subscriber.
-            if !channel.subscribers.contains(&id).await? {
-                // Send the latest message if any.
-                if let Some(latest_height) = channel.block_height.get() {
-                    let target = Target::channel(id, full_name.clone());
-                    let mut outbox = self.outboxes.try_load_entry_mut(&target).await?;
-                    if outbox.schedule_message(*latest_height)? {
-                        *outbox_counters.entry(*latest_height).or_default() += 1;
+        let stream = raw_result.subscribe.into_iter().zip(channels);
+        let stream = stream::iter(stream)
+            .map(|((name, id), mut channel)| async move {
+                let mut result = None;
+                let full_name = ChannelFullName {
+                    application_id,
+                    name,
+                };
+                // Add subscriber.
+                if !channel.subscribers.contains(&id).await? {
+                    // Send the latest message if any.
+                    if let Some(latest_height) = channel.block_height.get() {
+                        let target = Target::channel(id, full_name.clone());
+                        result = Some((target, *latest_height));
                     }
+                    channel.subscribers.insert(&id)?;
                 }
-                channel.subscribers.insert(&id)?;
+                Ok::<Option<(Target, BlockHeight)>, ChainError>(result)
+            })
+            .buffer_unordered(C::MAX_CONNECTIONS);
+        let infos = stream.try_collect::<Vec<_>>().await?;
+        let mut targets = Vec::new();
+        let mut heights = Vec::new();
+        for (target, height) in infos.into_iter().flatten() {
+            targets.push(target);
+            heights.push(height);
+        }
```

Comment:
I think this is `Iterator::unzip`:

```suggestion
        let (targets, heights): (Vec<_>, Vec<_>) = infos.into_iter().flatten().unzip();
```